src/App.tsx

```diff
@@ -15,49 +17,118 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  NONE,
+  ALPHABET,
+  LENGTH,
+}
+
+type ReorderOptions = {
+  sortType: SortType,
+  isReversed: boolean,
+};
+
+export function getReorderedGoods(
+  goods: string[],
+  { sortType, isReversed }: ReorderOptions,
+) {
+  const visibleGoods = [...goods];
+
+  visibleGoods.sort((previousGood, currentGood) => {
+    switch (sortType) {
+      case SortType.ALPHABET:
+        return previousGood.localeCompare(currentGood);
+
+      case SortType.LENGTH:
+        return previousGood.length - currentGood.length;
+
+      default:
+        return 0;
+    }
+  });
+
+  return isReversed
+    ? visibleGoods.reverse()
+    : visibleGoods;
+}
+
 export const App: React.FC = () => {
+  const [isReversed, setReversed] = useState(false);
+  const [sortType, setSortByType] = useState(SortType.NONE);
+  const isVisibleReset = isReversed || sortType !== SortType.NONE;
+  const goods = getReorderedGoods(goodsFromServer, {
+    isReversed,
+    sortType,
+  });
+
+  const reverseList = () => {
+    setReversed(!isReversed);
```

Comment:
In ReactJS, it’s recommended to use functional `setState` instead of the regular `setState`. The reason for this is that the functional form of `setState` can help avoid issues with stale state, and ensure that the state is updated correctly.
Here are some rules to consider when using functional `setState` in React:
1. **Always use functional `setState` when updating state based on previous state:** If you need to update the state based on its previous value, it’s recommended to use functional `setState` to ensure that the state is updated correctly. Here’s an example:
```jsx
// using functional setState
this.setState(prevState => ({ count: prevState.count + 1 }));
```
2. **Use regular `setState` for simple updates:** If you’re updating the state with a simple value, such as a boolean or a string, you can use regular `setState`. Here’s an example:
```jsx
// using regular setState
this.setState({ isModalOpen: true });
```
3. **Use functional `setState` for complex state updates:** If you’re updating the state with a complex value, such as an array or an object, it’s recommended to use functional `setState` to ensure that the state is updated correctly. Here’s an example:
```jsx
// using functional setState with object spread syntax
this.setState(prevState => ({ ...prevState, items: newItems }));
```
4. **Be careful with nested state updates:** If you have nested state updates, it’s important to use functional `setState` to ensure that the state is updated correctly. Here’s an example:
```jsx
// incorrect use of setState with nested state updates
this.setState({ items: [] });
this.setState(prevState => ({ items: [...prevState.items, newItem] }));
// correct use of functional setState with nested state updates
this.setState(prevState => ({ items: [...prevState.items, newItem] }));
```
In summary, using functional `setState` in ReactJS can help avoid issues with stale state and ensure that the state is updated correctly. Always use functional `setState` when updating state based on its previous value, and be careful with nested state updates.